src/application/utilities/github.ts

```diff
@@ -0,0 +1,29 @@
+import type { Brand } from "../types/utils";
+
+const SNAPSHOT_MATCHER = /^0.0.0-pr-(?<prNumber>\d+)-(?<timestamp>\d+)$/;
+const TIMESTAMP_MATCHER =
+  /^(?<year>\d{4})(?<month>\d{2})(?<day>\d{2})(?<hour>\d{2})(?<minutes>\d{2})(?<seconds>\d{2})$/;
+
+export type SnapshotVersion = Brand<string, "SnapshotVersion">;
```

Comment:
I'll give you a link to [this article](https://egghead.io/blog/using-branded-types-in-typescript) which I think does a lot better job of explaining the usefulness of a branded type.

With that, I'll explain the _why_ this is useful in this situation.

There are a few utility functions in this file where you'll see it shine, namely the `parseSnapshotRelease` function and the `parseSnapshotTimestamp` function. These functions are meant to operate on snapshot version strings because it allows me to pull out specific information from the version string.

Let's look at this in the context of what this would look like if `version` were just a string type:

```ts
function parseSnapshotRelease(version: string) {
  const match = SNAPSHOT_MATCHER.exec(version);

  // ...
}
``` 

Because this operates on a plain `string` type, I can't make any guarantees about its format since it could be an actual published version (i.e. `3.10.0`) or a random set of characters. 

In other words, this is perfectly valid to TypeScript:

```ts
parseSnapshotVersion('abcdefg')
```

Because of this, my `match` might fail and return `null`, which forces me to handle this in my function. To guard against this, I can either `throw` to indicate we got an unexpected string (yuck, now I have to add error handling everywhere) or return `null` (slightly better, but now I have to do non-null assertions or ternary functions in other places). 

Here is an implementation of the helper function that handles cases of invalid snapshot version strings and returns `null`:

```ts
function parseSnapshotRelease(version: string) {
  const match = SNAPSHOT_MATCHER.exec(version);

  return match && match.groups ? {
    prNumber: match.groups.prNumber,
    timestamp: match.groups.timestamp
  } : null
}
```

Unfortunately this mean the `null` can cascade through my code. This is especially infuriating if we have to handle those `null` values after I've checked the format using the `isSnapshotVersion` helper function:

```ts
if (isSnaphotVersion(version)) {
  const details = parseSnapshotRelease(version);
//      ^? { prNumber: string; timestamp: string } | null

// we just checked if we have a valid snapshot version string, but I'm still getting `null` back!
}
```

There are two ways I can handle this:

1. Handle `null` everywhere I use `details` to make TypeScript happy
2. Add a non-null assertion (`!`) to tell TypeScript that we know better.

While 2. is the solution that we'd likely use and be ok with, it would be nice if we can avoid the non-null assertion to begin with (preferred) while also avoiding the extra runtime code to handle `null` everywhere.

This is where the branded type comes in. While `version` is just a `string` type, the branded type allows me to tell TypeScript that its a _specific type_ of string. And because its a type, I can update my helper functions to only allow strings that are valid snapshot versions, which forces me to ensure I've checked the string ahead of time.

This adds the extra layer of type safety to my code ensuring that I've actually performed a check to ensure the `version` is not some random string. That happens by making the `isSnapshotVersion` a predicate function:

```ts
function isSnapshotVersion(version: string): version is SnapshotVersion
```

Because we "cast" that type, we now have better type safety in areas where I know I'm working with a snapshot version string (and that type follows the variable around!)

Revisiting `parseSnapshotRelease`, now I can ask it only to perform work on snapshot version strings:

```ts
function parseSnapshotRelease(version: SnapshotVersion)
```

which means I can make guarantees about its format when I match it against the regex to pull out specific information:

```ts
export function parseSnapshotRelease(version: SnapshotVersion) {
  const match = SNAPSHOT_MATCHER.exec(version)!;

  // No more nulls! We can safely use `!` because we've already guaranteed its format
  return {
    prNumber: match.groups!.prNumber,
    timestamp: match.groups!.timestamp,
  };
}
```

And now its much easier to guarantee type safety:
```ts
const version: string = '...';

// Error! We haven't guaranteed version is a snapshot version
parseSnapshotVersion(version) 

if (isSnapshotVersion(version)) {
  // now `version` is a `SnapshotVersion` type and we can use this function.
  // no more nulls!
  const details = parseSnapshotRelease(version);
//  ^? { prNumber: string; timestamp: string }
}
```

Hope this helps!

Reply:
Makes a ton of sense, thanks for the write-up @jerelmiller!

I find the way you've used branded types with type guards to be more ergonomic than the custom validation example in the post you shared (where validation is done inline in a fn and only at the end the result is cast to the branded type), but it was a great read in general. Thanks for sending!

Reply:
I really need to spin up my own blog so I can write about these things 😂